### DISCUSSION TITLE
Tendon driven motor actuators don't work

### DISCUSSION THREAD

**Discussion by Infraviored** [google-deepmind/mujoco#74] [Asking for Help] [answered]
I am trying to simulate a tendon-driven humanoid robot in mujoco.

when using 
`<muscle name="motor0" tendon="motor0" />`
the motor is controllable

using 
`<motor name="motor0" tendon="motor0" />`
the motor stays fixed for all control inputs in [-1,1]

even when using
`<cylinder name="motor0" tendon="motor17" forcerange="0 10000" ctrlrange="0 10000" lengthrange="0 10000" />`
or replacing motor with position, velocity or cylinder there is no effect of contro input.

Help would be greatly appreciated

**Comment by yuvaltassa** [comment]
Hi,
If you make a minimal model (e.g. one joint and one tendon) and post here, we will help you debug.

**Reply by Infraviored** [reply]
okay i did. Running this with 
cd ~/.mujoco && ./mujoco200/bin/simulate /src/minimal.xml 
if of relevance.

**Comment by Infraviored** [comment]
Created a minimal model.
red tendon is attached to a motor, green one to a muscle. Green can be controlled, red can't 

``<mujoco model="test">
    <compiler coordinate="global"/>

    <default>
        <geom rgba=".9 .7 .1 1" size="0.01"/>
        <site type="sphere" rgba=".9 .9 .9 1" size="0.005"/>
        <joint type="hinge" axis="0 1 0" limited="true" range="-30 60" solimplimit="0.95 0.95 0.1"/>
    </default>

    <visual>
        <headlight diffuse=".7 .7 .7"/>
    </visual>

    <worldbody>
        <site name="s0" pos="-0.03 0.01 0.4"/>
        <site name="s1" pos="-0.03 -0.01 0.4"/>

      <body>
      <geom type="capsule" fromto="0 0 0.3  0.2 0 0.3"/>
            <geom name="g1" type="cylinder" fromto="0.0 0.015 0.3  0.0 -0.015 0.3"
                    size="0.02" rgba=".3 .9 .3 .4"/>
        <joint pos="0 0 0.3"/>
            <site name="s2" pos="0.1 0 0.31"/>
      </body>
    </worldbody>

    <tendon>
        <spatial name="red" width="0.002" rgba=".95 .3 .3 1" limited="true" range="0 0.33">
            <site site="s0"/>
            <geom geom="g1"/>
            <site site="s2"/>
        </spatial>
        <spatial name="green" width="0.002" rgba=".3 .95 .3 1" limited="true" range="0 0.33">
            <site site="s1"/>
            <geom geom="g1"/>
            <site site="s2"/>
        </spatial>
    </tendon>
    <actuator>
        <motor name="motor" tendon="red"/>
        <muscle name="muscle" tendon="green"/>
    </actuator>
</mujoco>
``

**Comment by yuvaltassa** [comment, marked as answer]
Your motor works just fine, it's just a bit weak so cannot pull your thing due to gravity.
Turn off gravity (in Disable Flags under the Physics tab) and you'll see that your motor works.
To increase the strength of your motor, increase the `gear` attribute to be bigger than 1. 
```
<motor name="motor" tendon="red" gear="10"/>
```

**Reply by Infraviored** [reply]
Got this working before my team meeting thanks to your help. Many thanks 

i still do not understand two things
-why do the motors contract when applying negative control inputs (muscles contract for positive inputs)
-What is the physical property being controlled? I have now some unknown force*10(gear) projected onto [-1,1]

**Comment by yuvaltassa** [comment]
Muscles can only contract, because that's how muscles work. Positive input corresponds to efferent excitation. It could have been negative, it's just a convention.

Your motors are applying a force (implicitly in units of Newton) to the tendon, in the range [-10, 10].

**Reply by Infraviored** [reply]
Thank you again. 
I built another demo (block of mass 1 on vertical sliding joint)  to show to my team that a gear of 10 is exactly what is needed to barely move the block (F=-9.81N) 
I still do not understand forcerange. From the documentation i expect forcerange="-10 10" be equivalent to gear="10" but it is not. Forcerange has no effect yet.

**Reply by yuvaltassa** [reply]
`forcerange` is just a clamping operator applied to the actuator output, just like `ctrlrange` is a clamping operator applied to the input. If e.g. you have `gear="10"` and `forcerange = "-10 10"`, then for controls in [-1, 1] the force will change between -10 and 10 and outside [-1, 1] it will clamp to [-10, 10]. As explained [here](https://mujoco.readthedocs.io/en/latest/computation.html#actuation-model), the output of your actuator is in `data.actuator_force`.

**Comment by GraceNiyo** [comment]
Hello,
I have a follow-up on this question. I have a similar problem. I created a tendon attached to a box (ceiling) and a ball. The ball has one hinge joint. The actuator is a muscle and the muscle but the is muscle in uncontrollable.  Here is my sample code. Any help is greatly appreciated. 
<mujoco model = "BallAndTendon">
   <compiler angle="degree"/>
    <option timestep="0.01" iterations="50" solver="Newton" integrator = "RK4" tolerance="1e-10" gravity="0 0 0"/> 
   <default>
        <joint limited="true"></joint>
        <muscle ctrllimited="true" ctrlrange="0 1"></muscle>
    </default>

    <worldbody>
        <light diffuse=".5 .5 .5" pos="0 0 3" dir="0 0 -1"/>
        <geom type="plane" size="1 1 0.1" rgba=".9 0 0 1"/>

        <body name ="ceiling" pos = "0 0 0.4">
            <geom type= "box" size ="0.02 0.02 0.001" mass ="0.1"></geom>
            <site name="C_B" type = "sphere" pos = "0 0 0"/>
        </body>

        <body name= "ball" pos= "0 0 0.3">
          <geom type ="sphere" size = " 0.01" mass = "1" rgba ="0 1 1 1" ></geom>
          <site name="B_C" type = "sphere" pos = "0 0 0"/>
          <joint type = "hinge"  axis = "1 1 1" range = "-45 45"/>
         </body>
    </worldbody>
    <tendon>
        <spatial name = "tendon1" springlength="0.008" width="0.001" rgba ="1 1 1 1">
          <site  site= "B_C"> </site>
          <site site= "C_B"> </site>
        </spatial>
    </tendon>
    <actuator>
     <muscle name="muscle1" tendon="tendon1" lengthrange="0.001 1.6" force = "50" ></muscle>
    </actuator>
</mujoco>

**Reply by yuvaltassa** [reply]
I didn't load this, but I don't see how pulling on the tendon could achieve anything.

The hinge joint is at 0 0 0 and the site that you're pulling on is also at 0 0 0, how can it create motion with no moment arm? Did you mean to use `<joint type="slide"/>`?

**Comment by GraceNiyo** [comment]
Hi,
Yes. The joint type is supposed to be  "slide."
Best,
Niyo
